### PR TITLE
Drivers: Intel: DMIC: Fix FW panic with multiple DAI capture, pausing, stopping and resuming

### PR DESCRIPTION
**Pull request by singalsu**

### CODE REVIEW

File: src/drivers/intel/dmic.c [hunk 1]
```diff
@@ -1649,9 +1656,9 @@ static int dmic_remove(struct dai *dai)
...
 	/* The next end tasks must be passed if another DAI FIFO still runs */
-	dai_info(dai, "dmic_remove(), dmic_active_fifos_mask = 0x%x",
-		 uncached_dmic_active_fifos_mask);
-	if (uncached_dmic_active_fifos_mask)
+	dai_info(dai, "dmic_remove(), dmic_active_fifos_mask = 0x%x, dmic_pause_mask = 0x%x",
+		 uncached_dmic_active_fifos_mask, uncached_dmic_pause_mask);
+	if (uncached_dmic_active_fifos_mask || uncached_dmic_pause_mask)
 		return 0;
 
 	pm_runtime_put_sync(DMIC_CLK, dai->index);
```
Comment (kv2019i) [inline src/drivers/intel/dmic.c:1664]: This is really problematic. There's no actual reference counting in cavs_pm_runtime_dis_dmic_clk_gating() and the "dai->index" passed to pm_runtime_put_sync() is just printed out.

The problem is that if we switch to pm_runtime_put_sync() that actual does refcounting, this code will break. FYI @lgirdwood @lyakh @marc-hb 

Oh well, this goes beyond this PR's scope.

**Review by kv2019i — APPROVED** [review]
Some notes inline, but actual patch content looks good!

File: src/drivers/intel/dmic.c [hunk 2]
```diff
@@ -136,7 +136,7 @@ static const uint32_t coef_base_b[4] = {PDM0_COEFFICIENT_B, PDM1_COEFFICIENT_B,
 
 /* Global configuration request for DMIC, need to use uncached address to access them */
 static SHARED_DATA struct sof_ipc_dai_dmic_params *dmic_prm[DMIC_HW_FIFOS];
-static SHARED_DATA int dmic_active_fifos_mask;
+static SHARED_DATA uint32_t dmic_active_fifos_mask;
```
[the remark this answers is not in the document]
Comment (singalsu) [reply src/drivers/intel/dmic.c:139]: Yep, that's the purpose, 16 bit int devices are not known in SOF plus new usage is unsigned. I added the new uint32_t pause_mask next to it so the active_fifos_mask looked weird. This was made a separe commit.

File: src/drivers/intel/dmic.c (hunk 1 above)
Comment (lgirdwood) [reply src/drivers/intel/dmic.c:1664]: Lets get the ref counting in the PM code. This PM code should also be able to transfer to Zephyr with minimal refactoring.

Comment (lgirdwood) [reply src/drivers/intel/dmic.c:1664]: it looks like we may need some locking around the mask manipulations .

**Review by ranj063 — APPROVED** [review]

**Comment by singalsu** [comment]
I just updated the PR: Added pause_mask clear to dmic_stop() in case it was called to stop without pause. A typo was fixed. Also added comments about spinlocks applied. Thanks for help with these @ranj063!

File: src/drivers/intel/dmic.c (hunk 1 above)
Comment (singalsu) [reply src/drivers/intel/dmic.c:1674]: I added 3rd commit about spinlocks. I was testing spinlock add in dmic_remove() but it froze. It turned out that dai_put() had already applied spinlock. I also found out that the registers write function was already covered by spinlock in set_config() so the patch adds only comments about spinlocks applied.

**Comment by ranj063** [comment]
looks good. Thanks @singalsu

**Review by lgirdwood — APPROVED** [review]
Lets fix the PM reference counting incrementally.

**Pull request merged**